pkg/blockbuilder/tsdb_test.go

```diff
@@ -170,28 +165,48 @@ func TestTSDBBuilder(t *testing.T) {
 				expSamples = expSamples[:0]
 				// B. This goes in this block.
-				addFloatSample(builder, lastEnd+100, lastEnd, currEnd, true, true)
+				addFloatSample(builder, lastEnd+100, 1, lastEnd, currEnd, true, true)
 				// C. This sample should be processed in the future.
-				addFloatSample(builder, currEnd+1, lastEnd, currEnd, true, false)
+				addFloatSample(builder, currEnd+1, 1, lastEnd, currEnd, true, false)
 
 				// 2. Processing records that were not processed before.
 				// A. Sample that belonged to previous processing period but came in late. Processed in current cycle.
-				addFloatSample(builder, lastEnd-5, lastEnd, currEnd, false, true)
+				addFloatSample(builder, lastEnd-5, 1, lastEnd, currEnd, false, true)
 				// B. Sample that belongs to the current processing period.
-				addFloatSample(builder, lastEnd+200, lastEnd, currEnd, false, true)
-				// C. This sample should be processed in the future.
-				addFloatSample(builder, currEnd+2, lastEnd, currEnd, false, false)
+				addFloatSample(builder, lastEnd+200, 1, lastEnd, currEnd, false, true)
+				// C. Sample that belongs to the current processing period but is a duplicate with different value.
+				addFloatSample(builder, lastEnd+200, 2, lastEnd, currEnd, false, true)
+				// The request is accepted, but its sample won't end up in the DB due to soft "ErrDuplicateSampleForTimestamp".
+				expSamples = expSamples[:len(expSamples)-1]
+				// D. This sample should be processed in the future.
+				addFloatSample(builder, currEnd+2, 1, lastEnd, currEnd, false, false)
+				// E. This sample is too old (soft error).
+				addFloatSample(builder, 0, 1, lastEnd, currEnd, false, true)
+				expSamples = expSamples[:len(expSamples)-1]
 
 				// 3. Out of order sample in a new record.
 				// A. In the current range but out of order w.r.t. the previous sample.
-				addFloatSample(builder, lastEnd+20, lastEnd, currEnd, false, true)
+				addFloatSample(builder, lastEnd+20, 1, lastEnd, currEnd, false, true)
 				// B. Before current range and out of order w.r.t. the previous sample. Already covered above, but this
 				// exists to explicitly state the case.
-				addFloatSample(builder, lastEnd-20, lastEnd, currEnd, false, true)
+				addFloatSample(builder, lastEnd-20, 1, lastEnd, currEnd, false, true)
+
+				// 4. Out of order sample with no OOO window configured.
+				oooUserID := "test-ooo-tenant"
+				rec := createRequest(oooUserID, floatSample(lastEnd+20, 1), nil, false)
+				allProcessed, err := builder.Process(context.Background(), rec, lastEnd, currEnd, false)
+				require.NoError(t, err)
+				require.True(t, allProcessed)
+
+				// A. "ErrOutOfOrderSample" (soft error)
+				rec = createRequest(oooUserID, floatSample(lastEnd-20, 1), nil, false)
+				allProcessed, err = builder.Process(context.Background(), rec, lastEnd, currEnd, false)
+				require.NoError(t, err)
+				require.True(t, allProcessed)
```

Comment:
To actually verify that it took the first sample and not the second, we should query the `test-ooo-tenant`'s block after it is created and expect only `lastEnd+20`.